api-authentication/server/index.js

```diff
@@ -0,0 +1,112 @@
+require('dotenv/config');
+const pg = require('pg');
+const argon2 = require('argon2'); // eslint-disable-line
+const express = require('express');
+const jwt = require('jsonwebtoken'); // eslint-disable-line
+const ClientError = require('./client-error');
+const errorMiddleware = require('./error-middleware');
+
+const db = new pg.Pool({
+  // eslint-disable-line
+  connectionString: process.env.DATABASE_URL,
+  ssl: {
+    rejectUnauthorized: false
+  }
+});
+
+const app = express();
+
+const jsonMiddleware = express.json();
+
+app.use(jsonMiddleware);
+
+app.post('/api/auth/sign-up', (req, res, next) => {
+  const { username, password } = req.body;
+  if (!username || !password) {
+    throw new ClientError(400, 'username and password are required fields');
+  }
+  argon2
+    .hash(password)
+    .then(hashedPassword => {
+      const sql = `
+        insert into "users" ("username", "hashedPassword")
+        values ($1, $2)
+        returning "userId", "username", "createdAt"
+      `;
+      const params = [username, hashedPassword];
+      return db.query(sql, params);
+    })
+    .then(result => {
+      const [user] = result.rows;
+      res.status(201).json(user);
+    })
+    .catch(err => next(err));
+});
+
+app.post('/api/auth/sign-in', (req, res, next) => {
+  const { username, password } = req.body;
+  if (!username || !password) {
+    throw new ClientError(401, 'invalid login');
+  }
+
+  /* your code starts here */
+
+  /**
+   * Query the database to find the "userId" and "hashedPassword" for the "username".
+   * Then, 😉
+   *    If no user is found,
+   *      throw a 401: 'invalid login' error.
+   *    If a user is found,
+   *      confirm that the password included in the request body matches the "hashedPassword" with `argon2.verify()`
+   *      Then, 😉
+   *        If the password does not match,
+   *          throw a 401: 'invalid login' error.
+   *        If the password does match,
+   *          Create a payload object containing the user's "userId" and "username".
+   *          Create a new signed token with `jwt.sign()`, using the payload and your TOKEN_SECRET
+   *          Send the client a 200 response containing the payload and the token.
+   *      Catch any error.
+   * Catch any error.
+   */
+
+  const sql =
+    'SELECT "userId", "username", "hashedPassword" FROM "users" WHERE "username" = $1;';
+  const params = [username];
+  db.query(sql, params).then(result => {
+    if (!result.rows[0]) {
+      throw new ClientError(401, 'Invalid Login');
+    }
+    const username = result.rows[0].username;
+    if (!username) {
+      throw new ClientError(401, 'invalid username');
+    } else if (username) {
```

Comment:
This doesn't need to be here, if there is a result from the database it will have a username, so there is no need to check that here. The check on line 76 will catch any bad usernames

Reply:
true! I just really love saving things in variables!
![Screenshot 2022-11-03 at 2 03 55 PM](https://user-images.githubusercontent.com/99847558/199834120-ecf48f05-d728-4d46-974e-82892cc7a8bb.png)